DataOnion/Auth/TwoFactorAuthService.cs

```diff
@@ -0,0 +1,259 @@
+    {
+        Task<RegisterDidResult> RegisterDidAsync(
+            RegisterDidParams parameters
+        );
+
+        Task<TwoFactorRequest?> FetchRequestAsync(
+            FetchRequestParams parameters
+        );
+
+        Task DeleteRequestsAsync(
+            string userId
+        );
+    }
+
+    public class TwoFactorAuthService : ITwoFactorAuthService
+    {
+        private readonly ITwoFactorRedisContext _redisContext;
+        private readonly ILogger? _logger;
+
+        public TwoFactorAuthService(
+            ITwoFactorRedisContext redisContext,
+            ILogger<TwoFactorAuthService>? logger = null
+        )
+        {
+            _redisContext = redisContext;
+            _logger = logger;
+        }
+
+        public async Task<RegisterDidResult> RegisterDidAsync(
+            RegisterDidParams parameters
+        )
+        {
+            var did = parameters.Did;
+            var method = parameters.Method;
+            var userId = parameters.UserId;
+            var code = parameters.VerificationCode;
+            var throttleTimeout = parameters.ThrottleTimeout;
+
+            bool tooManyRequests = false;
+            TwoFactorRequest? newRequest = null;
+            int timeoutRemaining = 0;
+
+            switch (method)
+            {
+                case VerificationMethod.Call:
+                    var recentCallRequests = await _redisContext.FetchTwoFactorRequestsAsync(
+                        userId,
+                        VerificationMethod.Call
+                    );
+
+                    var recentCallRequest = recentCallRequests.FirstOrDefault();
+                    var callRequestValidity = CheckTwoFactorTimeout(recentCallRequest, throttleTimeout);
+
+                    if (recentCallRequest != null && !callRequestValidity.Item1)
+                    {
+                        tooManyRequests = true;
+                        timeoutRemaining = callRequestValidity.Item2;
+                    }
+                    break;
+                case VerificationMethod.Text:
+                    var recentTextRequests = await _redisContext.FetchTwoFactorRequestsAsync(
+                        userId,
+                        VerificationMethod.Text
+                    );
+
+                    var recentTextRequest = recentTextRequests.FirstOrDefault();
+                    var textRequestValidity = CheckTwoFactorTimeout(recentTextRequest, throttleTimeout);
+
+                    if (recentTextRequest != null && !textRequestValidity.Item1)
+                    {
+                        tooManyRequests = true;
+                        timeoutRemaining = textRequestValidity.Item2;
+                    }
+                    break;
+                // In the case that we ever add another form of 2FA I don't
+                // want this implementation to prevent it from working if
+                // we forget to update this.
+                default:
+                    break;
```

Comment:
We can't throw an exception from DataOnion because the error handler won't catch it, and the API response will just be a big ugly stack trace.

One option is to extend `RegisterDidResult` to include whether the method was invalid. We could do that by changing `AreThereTooManyRequests` to something like `IsThereAnException` and add a new property to hold the error code, or we could replace `AreThereTooManyRequests` with an enum like `RegisterDidResultType`.

Alternatively, we could remove `VerificationMethod.Unknown` entirely and force the client to deal with it before they call DataOnion. I'm leaning towards this option.

Reply:
You can't throw an exception from DataOnion? Why?